lahja/trio/endpoint.py

```diff
@@ -0,0 +1,905 @@
+        await self._message_send_channel.aclose()
+        await self._connection_send_channel.aclose()
+        await self._broadcast_send_channel.aclose()
+
+        del self._message_send_channel
+        del self._connection_send_channel
+        del self._broadcast_send_channel
+
+        self.logger.debug("%s: stopped", self)
+
+    #
+    # Background processes
+    #
+    async def _process_broadcasts(
+        self,
+        channel: trio.abc.ReceiveChannel[ConcreteBroadcast],
+        nursery: trio_typing.Nursery,
+    ) -> None:
+        self._process_broadcasts_running.set()
+        async for (item, config) in channel:
+            nursery.start_soon(self.broadcast, item, config)
+
+    async def _process_connections(
+        self,
+        connection_receive_channel: trio.abc.ReceiveChannel[ConnectionConfig],
+        nursery: trio_typing.Nursery,
+    ) -> None:
+        """
+        Long running process that establishes connections to endpoint servers
+        and runs the handler for receiving events sent by the server over that
+        connection.
+        """
+        self.logger.debug("%s: starting new connection channel", self)
+        self._connection_loop_running.set()
+        async for config in connection_receive_channel:
+            # Allow some time for for the IPC socket to appear
+            with trio.fail_after(constants.IPC_WAIT_SECONDS):
+                await _wait_for_path(trio.Path(config.path))
+
+            # Establish the socket connection
+            connection = await TrioConnection.connect_to(config.path)
+
+            # Create the remote
+            remote = TrioRemoteEndpoint(
+                self.name,
+                connection,
+                self._remote_subscriptions_changed,
+                self._message_send_channel.send,
+            )
+            nursery.start_soon(self._run_remote_endpoint, remote)
+
+    async def _process_received_messages(
+        self,
+        receive_channel: trio.abc.ReceiveChannel[Broadcast],
+        nursery: trio_typing.Nursery,
+    ) -> None:
+        """
+        Consume events that have been received from a remote endpoint and
+        process them.
+        """
+        self._message_processing_loop_running.set()
+        async for (item, config) in receive_channel:
+            event = self._decompress_event(item)
+            nursery.start_soon(self._process_item, event, config)
+
+    async def _monitor_subscription_changes(self) -> None:
+        while not self.is_stopped:
+            # We wait for the event to change and then immediately replace it
+            # with a new event.  This **must** occur before any additional
+            # `await` calls to ensure that any *new* changes to the
+            # subscriptions end up operating on the *new* event and will be
+            # picked up in the next iteration of the loop.
+            await self._subscriptions_changed.wait()
+            self._subscriptions_changed = trio.Event()
```

Comment:
Now that we're in trio world this might be easier to implement with a channel. Using a channel would also allow for the batching of events that you wanted, when an event comes in this can first empty the queue and then send messages to the remotes.

Reply:
I think you're right but it feels roughly equivalent and I'm inclined to leave this as-is.